### PR TITLE
[SYCL-MLIR] Detect default `sycl.id` constructors

### PR DESCRIPTION
**Pull request by victor-eds**
Detect default construction of `sycl.id` both checking whether all of the stored values are a constant 0 or the destination alloca is being memset to zero.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 1]
```diff
@@ -665,24 +691,6 @@ class RaiseArrayConstructorBasePattern
   }
 
 private:
-  std::optional<LLVM::LLVMArrayType>
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:668]: moved above

File: polygeist/tools/cgeist/Test/Verification/sycl/raising/failing_host_raising_id_constructors.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
+// RUN: clang++ -O1 %s -S -emit-mlir -o - -fsycl -fsycl-raise-host -Xclang -opaque-pointers -w | FileCheck %s
+// RUN: clang++ -O2 %s -S -emit-mlir -o - -fsycl -fsycl-raise-host -Xclang -opaque-pointers -w | FileCheck %s
+// RUN: clang++ -O3 %s -S -emit-mlir -o - -fsycl -fsycl-raise-host -Xclang -opaque-pointers -w | FileCheck %s
+// XFAIL: *
+
+// COM: Copy and move constructors not being detected yet for Dimensions > 2.
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/sycl/raising/failing_host_raising_id_constructors.cpp:6]: I will address these in a future PR

**Review by sommerlukas — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 3]
```diff
@@ -240,6 +240,22 @@ static Value getHandler(FunctionOpInterface op) {
...
+
+  auto detailType = dyn_cast<LLVM::LLVMStructType>(structTy.getBody().front());
+  static llvm::Regex arrayRegex{"class.sycl::_V1::detail::array(\\.[0-9]+)?"};
+  if (!detailType || !arrayRegex.match(detailType.getName()) ||
+      detailType.getBody().empty())
+    return std::nullopt;
+
+  auto arrayType = dyn_cast<LLVM::LLVMArrayType>(detailType.getBody().front());
+
+  return arrayType;
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:256]:
```suggestion
  return dyn_cast<LLVM::LLVMArrayType>(detailType.getBody().front());
```

**Review by etiotto — APPROVED** [review]

File: polygeist/test/polygeist-opt/sycl/sycl-raise-host.mlir [hunk 4]
```diff
@@ -204,6 +204,61 @@ llvm.func @raise_id_3() -> !llvm.ptr {
 
 llvm.func @_Z6numberv() -> (i64)
 
+// COM: We should not find this pattern with dimensions == 1
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/sycl/sycl-raise-host.mlir:207]: icic.

**Review by whitneywhtsang — APPROVED** [review]

**Comment by victor-eds** [comment]
Failure unrelated to this test.

**Pull request merged**